### PR TITLE
[Feature] add clone & download repo button in programming details view 

### LINKED ISSUE #2552: Clone and download zip button for instructor repositories (exercise, test, solution)
We offer a clone button for students.
It would be great if we could offer a clone button for instructors as well for the exercise, test and solution repos.
In addition, a download zip button would be nice for the exercise, test and solution repos for cases in which the instructor wants to share single repos with someone or only wants to try out the exercise locally without the need to change something (using git commit + push). This could be done with the existing functionality to export repos (using a git clone and zip operation on the Artemis server in the repos-download folder)

This should be implemented on the programming exercise view page, e.g. on https://artemis.ase.in.tum.de/course-management/10/programming-exercises/1696 next to the respective URLs:

image

### PR DESCRIPTION
**Pull request by ivanchimeno**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

Resolves #2552 

It should be possible for instructors to clone and/or download a zipped version of the test, template, and solution repositories (separately) for programming exercises. 

### Description

The programming exercise details view displays the URLs for the template, test, and solution repositories. This PR adds a "clone repo" and "download repo" button next to each URL.

This PR does the following:
-  Extracts the existing clone repo button into its own `jhi-clone-repo-button` component so that it can be used in views other than the student view
-  Creates a new component `jhi-programming-exercise-instructor-repo-download` which enables instructors to download the repo as a zipped file
- Adds a new endpoint `/export-instructor-repository/{repositoryTypeString}` that enabled instructors to export a test, template, or solution repository 

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration -> Exercise
3. Click on a programming exercise and navigate to where the URLs are located
4. You should see the buttons
5. Test that the zipped files are the same of the respective repository
6. The "Clone Repo" functionality should behave the same way as it does for students

### Test Coverage

programming-exercise-detail.component.ts 50.47%
programming-exercise.service.ts 61.9%
programming-exercise-instructor-repo-download.component.ts 100%
exercise-details-student-actions.component.ts 77.05%
clone-repo-button.component.ts 94.44%
ProgrammingExerciseExportService.java 48.1% 
ProgrammingExerciseResource.java   87.1%
GitService.java  74.5%

### Screenshots

image
image

### CODE REVIEW

**Review by airelawaleria — DISMISSED** [review]
Worked for me on ts1 with all 3 repos

**Review by philgabr — APPROVED** [review]
Works as intended :+1:

**Review by simonzilker — APPROVED** [review]
Looks good!

**Review by JohannesWeiss — APPROVED** [review]
Tested on ts3, works as expected there.

**Review by marvinchimeka — DISMISSED** [review]
Works as intended.

**Review by sebastiantpeter — CHANGES_REQUESTED** [review]
Tested on ts1. Seems to work fine with one exception: Is it possible, that if you click "Download Repo" next to the Template Repository Url you actually download the Test Repository and vice versa?

**Review by DenizShabani — DISMISSED** [review]

**Comment by ivanchimeno** [comment]
> Tested on ts1. Seems to work fine with one exception: Is it possible, that if you click "Download Repo" next to the Template Repository Url you actually download the Test Repository and vice versa?

You're correct. Thanks for the catch!

**Review by EneaGore — COMMENTED** [review]
Noticed same thing as Sebastian but otherwise it worked fine.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java [hunk 1]
```diff
@@ -902,11 +906,54 @@ else if (!authCheckService.isAtLeastTeachingAssistantForExercise(programmingExer
...
+     * POST /programming-exercises/:exerciseId/export-instructor-respository/:repositoryTypeString : sends a test, solution or template repository as a zip file
+     * @param exerciseId The id of the programming exercise
+     * @param repositoryTypeString The type of repository to zip and send
+     * @return ResponseEntity with status
+     * @throws IOException if something during the zip process went wrong
+     */
+    @GetMapping(Endpoints.EXPORT_INSTRUCTOR_REPOSITORY)
+    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
+    @FeatureToggle(Feature.PROGRAMMING_EXERCISES)
+    public ResponseEntity<Resource> exportTestRepositoryForProgrammingExercise(@PathVariable long exerciseId, @PathVariable String repositoryTypeString) throws IOException {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:919]:
```suggestion
    public ResponseEntity<Resource> exportInstructorRepositoryForProgrammingExercise(@PathVariable long exerciseId, @PathVariable String repositoryTypeString) throws IOException {
```

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:919]: Also I think you can map the enum here directly, see e.g. https://www.baeldung.com/spring-enum-request-param

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseExportService.java [hunk 2]
```diff
@@ -94,11 +95,63 @@ public ProgrammingExerciseExportService(ProgrammingExerciseRepository programmin
...
     }
 
+    /**
+     * Exports a repository available for an instructor/tutor for a given programming exercise. This can be a template,
+     * solution, or tests repository
+     * @param exercise The programming exercise that has the repository
+     * @param repositoryType the type of repository to export
+     * @return a zipped file
+     */
+    public File exportInstructorRepositoryForExercise(ProgrammingExercise exercise, RepositoryType repositoryType) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseExportService.java:105]: I think there is quite some duplicated code here. Can you try to unify this with the other export functionality?

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java [hunk 3]
```diff
@@ -921,6 +921,29 @@ public Path zipRepository(Repository repo, String targetPath, boolean hideStuden
...
 
+    /**
+     * Zip the content of a git repository available for an instructor/tutor.
+     * @param repo Local Repository Object.
+     * @param zipName The name of the zipped file
+     * @param targetPath path where the repo is located on disk
+     * @return The path to the zipped file
+     * @throws IOException if the zipping process failed.
+     */
+    public Path zipInstructorRepository(Repository repo, String zipName, String targetPath) throws IOException {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:932]: do we really need this? or could this be implemented with an existing method?
We should avoid code duplication as it leads to additional effort

**Review by krusche — CHANGES_REQUESTED** [review]
I have some inline comments related to code duplication.
Also, the sever integration tests are missing, please write some and make sure that all new REST calls are covered and that at least 80% of the new server code is covered.

File: src/main/webapp/app/exercises/programming/shared/actions/programming-exercise-instructor-repo-download.component.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            [btnType]="ButtonType.INFO"
+            [btnSize]="ButtonSize.SMALL"
+            [shouldSubmit]="false"
+            [featureToggle]="FeatureToggle.PROGRAMMING_EXERCISES"
+            [icon]="'download'"
+            [title]="'artemisApp.programmingExercise.export.downloadRepo'"
+            (onClick)="exportRepository()"
+        ></jhi-button>
+    `,
+    styles: [],
```
Comment (iozbeyli) [inline src/main/webapp/app/exercises/programming/shared/actions/programming-exercise-instructor-repo-download.component.ts:22]: If there is no style, you can just remove this.

File: src/main/webapp/app/shared/components/clone-repo-button/clone-repo-button.component.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+import { TranslateService } from '@ngx-translate/core';
+import { ProfileInfo } from 'app/shared/layouts/profiles/profile-info.model';
+import { AccountService } from 'app/core/auth/account.service';
+import { User } from 'app/core/user/user.model';
+import { ProfileService } from 'app/shared/layouts/profiles/profile.service';
+
+@Component({
+    selector: 'jhi-clone-repo-button',
+    templateUrl: './clone-repo-button.component.html',
+    styles: [],
```
Comment (iozbeyli) [inline src/main/webapp/app/shared/components/clone-repo-button/clone-repo-button.component.ts:13]: Same here

File: src/test/javascript/spec/component/shared/clone-repo-button.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        component.sshTemplateUrl = 'ssh://git@bitbucket.ase.in.tum.de:7999/';
+        component.useSsh = true;
+
+        component.isTeamParticipation = true;
+        let url = component.getHtmlOrSshRepositoryUrl();
+        expect(url).to.equal('ssh://git@bitbucket.ase.in.tum.de:7999/ITCPLEASE1/itcplease1-exercise.git');
+
+        component.isTeamParticipation = false;
+        url = component.getHtmlOrSshRepositoryUrl();
+        expect(url).to.equal('ssh://git@bitbucket.ase.in.tum.de:7999/ITCPLEASE1/itcplease1-exercise.git');
```
Comment (iozbeyli) [inline src/test/javascript/spec/component/shared/clone-repo-button.component.spec.ts:129]: This part is repeated maybe move it into a function.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseExportService.java (hunk 2 above)
Comment (ivanchimeno) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseExportService.java:107]: I refactored a bit and hopefully the code is a bit better now.

**Review by philgabr — COMMENTED** [review]
The issue mentioned by @sebastiantpeter seems to be fixed :+1:

Everything works as intended.

**Review by JannisBittlmayer — APPROVED** [review]
Tested on TS1 with @FlorianStang, works as intended

**Review by FlorianStang — APPROVED** [review]
Tested on TS1 with @JannisBittlmayer, everything worked as intended. Nice changes

**Review by krusche — APPROVED** [review]

**Pull request merged**